language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -55,48 +124,112 @@ type VarEnv = UniqueMap TermUnique Closure
 data CekEnv = CekEnv
     { _cekEnvMeans  :: DynamicBuiltinNameMeanings
     , _cekEnvVarEnv :: VarEnv
+    , _cekEnvProtocolParameters :: ()
     }
 
+data ExBudget = ExBudget { _exBudgetCPU :: ExCPU, _exBudgetMemory :: ExMemory }
+
+-- TODO memory should probably be fixed size - but then how do we handle overflow?
+newtype ExMemory = ExMemory Integer -- Probably machine word size
+  deriving (Eq, Ord)
+  deriving newtype Num
+  deriving (Semigroup, Monoid) via (Sum Integer)
+newtype ExCPU = ExCPU Integer
+  deriving (Eq, Ord)
+  deriving newtype Num
+
+makeLenses ''ExBudget
+
 -- | The monad the CEK machine runs in.
-type CekM = ReaderT CekEnv (Either CekMachineException)
+type CekMBase w s = ExceptT CekMachineException (RWS CekEnv w s)
```

Comment:
The writer for tallying the costs, the state if you want to budget it.

Reply:
Better stack now with `ReaderT CekEnv (ExceptT CekMachineException (State (CekState w s)))`.